orchagent/neighorch.cpp

```diff
@@ -354,13 +354,31 @@ void NeighOrch::doTask(Consumer &consumer)
             if (m_syncdNeighbors.find(neighbor_entry) == m_syncdNeighbors.end() || m_syncdNeighbors[neighbor_entry] != mac_address)
             {
                 if (addNeighbor(neighbor_entry, mac_address))
+                {
                     it = consumer.m_toSync.erase(it);
+                }
                 else
+                {
                     it++;
+                    continue;
+                }
             }
             else
+            {
                 /* Duplicate entry */
                 it = consumer.m_toSync.erase(it);
+            }
+
+            /* Remove remaining DEL operation in m_toSync for the same neighbor.
+             * Since DEL operation is supposed to be executed before SET for the same neighbor
+             * A remaining DEL after the SET operation means the DEL operation failed previously and should not be executed anymore
+             */
+            auto it_rem = consumer.m_toSync.find(key);
```

Comment:
>find [](start = 44, length = 4)

Is it more efficient to travel backward and erase everything with the same key?

Reply:
Try [https://en.cppreference.com/w/cpp/iterator/make_reverse_iterator](https://en.cppreference.com/w/cpp/iterator/make_reverse_iterator) to make code cleaner?

---
In reply to: [512406052](https://github.com/Azure/sonic-swss/pull/1485#discussion_r512406052) [](ancestors = 512406052)

Reply:
actually I had a second thought here, since it is a multi-map, lookup with key should be fast.  To make it simple, how about have simple code as below: 

rm_num = consumer.m_toSync.erase(key);
if (rm_num) 
     SWSS_LOG_NOTICE(...);

above will handle key exist or not exist case.  

@qiluo-msft @shi-su

Reply:
@zhenggen-xu @qiluo-msft Thanks for the suggestions. I think erasing by key has a logarithm complexity. I updated the code to use a reverse iterator to do the job. In this scenario, since there would be at most one pending DEL operation, using the reverse iterator should achieve a constant complexity. Please let me know what you think.

Reply:
yeah, understood it was logarithm but that extra time did not seem to be a concern in our case, but it gives us a better readability and maintainability. 
Anyway, I don't have strong opinion here, either way is not complex anyways.   So I leave it to @qiluo-msft